parser/ParseIndexProperties.hpp

```diff
@@ -0,0 +1,407 @@
+   * @param reason_for_invalid_description Printable reason for invalid index.
+   * @param invalid_property_node Parse node that may be caused the index
+   *                              to become valid.
+   **/
+  IndexProperties() {
+  }
+
+  /**
+   * @brief Virtual Destructor
+   */
+  virtual ~IndexProperties() {
+  }
+
+  /**
+   * @brief Checks whether this instance describes a valid index or not.
+   *
+   * @return True if index description is valid, false otherwise.
+   **/
+  virtual bool isIndexDescriptionValid() const = 0;
+
+  /**
+   * @brief Returns the parse node corresponding for invalid index instance, if any.
+   *
+   * @return The parse node.
+   **/
+  virtual const ParseKeyValue* getInvalidPropertyNode() const = 0;
+
+  /**
+   * @brief Returns printable string that specifies reason for invalid index instance.
+   *
+   * @return The string describing the reason for invalid index
+   **/
+  virtual std::string getReasonForInvalidIndexDescription() const = 0;
+
+  /**
+   * @brief Returns a protobuf object representing the index description.
+   *
+   * @return An IndexSubBlockDescription object.
+   **/
+  const IndexSubBlockDescription* getIndexDescription() const {
+    return index_sub_block_description_.get();
+  }
+
+  /**
+   * @brief Modifies the index description given a set of custom properties
+   *        specified as a key-value list of properties.
+   *
+   * @param key_value_list Key Value list of custom properties.
+   * @return Returns true if the given set of custom properties was valid
+   *         and was applied correctly, otherwise returns false.
+   **/
+  virtual bool addCustomProperties(const PtrList<ParseKeyValue> *key_value_list) = 0;
+
+  /**
+   * @brief A helper function that checks whether the given list of custom properties
+   *        contains duplicate keys or not.
+   *
+   * @param key_value_list Key Value list of custom properties.
+   * @param invalid_node Double pointer to an invalid node that is used to store duplicate node, if any.
+   * @return Returns true if the given set of custom properties contains duplicate
+   *         keys, otherwise returns false.
+   **/
+  bool hasDuplicateCustomProperty(const PtrList<ParseKeyValue> *key_value_list, const ParseKeyValue **invalid_node) {
+    // Define an existence map, which checks for duplicate properties in the key_value_list.
+    std::unordered_map<std::string, bool> is_duplicate_property;
+    for (const ParseKeyValue &key_value : *key_value_list) {
+      const std::string key = ToLower(key_value.key()->value());
+      if (is_duplicate_property.find(key) != is_duplicate_property.end()) {
+        *invalid_node = &key_value;
+        return true;  // duplicate found, return true.
+      } else {
+        is_duplicate_property[key] = true;
+      }
+    }
+    return false;  // no duplicates, return false.
+  }
+
+ protected:
+  std::unique_ptr<IndexSubBlockDescription> index_sub_block_description_;
+};
```

Comment:
Do you intend to allow this class and its derived classes copyable? If not, add `DISALLOW_COPY_AND_ASSIGN(IndexProperties)` and do so for all derived classes.